dbt/adapters/factory.py

```diff
@@ -30,4 +35,28 @@ def get_adapter_by_name(adapter_name):
 
 
 def get_adapter(config):
-    return get_adapter_by_name(config.credentials.type)
+    adapter_name = config.credentials.type
+    if adapter_name in _ADAPTERS:
+        return _ADAPTERS[adapter_name]
+
+    adapter_type = get_adapter_class_by_name(adapter_name)
+    with _ADAPTER_LOCK:
+        # check again, in case something was setting it before
+        if adapter_name in _ADAPTERS:
+            return _ADAPTERS[adapter_name]
+
+        adapter = adapter_type(config)
+        _ADAPTERS[adapter_name] = adapter
+        return adapter
```

Comment:
very into this. since we are getting more into the habit of using static classes as namespaces, should this be `AdapterFactory.create()` rather than `dbt.adapters.factory.get_adapter()`?

this PR is big enough, I don't want you to add it to this. but worth considering for the future.

Reply:
I'm not sure. I think this is pretty reasonable to keep as a module-level entity, I don't think making the factory an object would gain much. The interface is pretty simple right now and I think that's very valuable. This kind of model in Python is pretty common (`logging.getLogger()` is the canonical example, I suppose).